src/main/java/com/ibm/cloud/cloudant/kafka/transforms/MapToStruct.java

```diff
@@ -0,0 +1,155 @@
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+
+    }
+
+    @Override
+    public SourceRecord apply(SourceRecord record) {
+        requireMapOrNull(record.value(), String.format("%s schema generation SMT", this.getClass().getName()));
+        if (record.value() != null) {
+            Map<String, Object> originalValueAsMap = (Map<String, Object>) record.value();
+            SchemaAndValue transformed = mapToStruct(originalValueAsMap);
+            return record.newRecord(record.topic(), record.kafkaPartition(), record.keySchema(), record.key(), transformed.schema(), transformed.value(), record.timestamp());
+        }
+        // If there is a null value we can't transform so return the original record
+        return record;
+    }
+
+    @Override
+    public ConfigDef config() {
+        return null;
+    }
+
+    @Override
+    public void close() {
+    }
+
+    static SchemaAndValue mapToStruct(Map<String, Object> jsonMap) {
+        // Collect in a TreeMap so the keys are ordered
+        Map<String, SchemaAndValue> intermediate = jsonMap.entrySet().stream().parallel()
+            .collect(Collectors.toMap(Map.Entry::getKey, MapToStruct::entryToSchemaAndValue, (v1, v2) -> v1, TreeMap::new));
+        SchemaBuilder schemaBuilder = SchemaBuilder.struct().optional();
+        intermediate.forEach((k, sav) -> {
+            schemaBuilder.field(k, sav.schema());
+        });
+        Schema schema = schemaBuilder.build();
+        Struct struct = new Struct(schema);
+        intermediate.forEach((k, sav) -> {
+            struct.put(k, sav.value());
+        });
+        return new SchemaAndValue(schema, struct);
+    }
+
+    static SchemaAndValue listToArray(List<Object> jsonArray) {
+        List<SchemaAndValue> array = jsonArray.stream()
+            .sequential()
+            .map(MapToStruct::objectToSchemaAndValue)
+            .collect(Collectors.toList());
+        List<Schema> schemas = array.stream().map(SchemaAndValue::schema).distinct().collect(Collectors.toList());
+        boolean hasNullValues = schemas.remove(NULL_VALUE_SCHEMA);
+        Schema arraySchema;
+        if( array.isEmpty() || (hasNullValues && schemas.isEmpty())) {
+            // There are certain arrays where we cannot infer the schema from the elements.
+            // If there are no elements to test, or if the only values are nulls.
+            // For these cases we just call it an optional string and move on.
+            // TODO log a warning, assume optional string
+            arraySchema = NULL_VALUE_SCHEMA;
+        } else if (schemas.size() > 1) {
+            // If there are multiple schemas then throw an exception because mixed type arrays are not supported by Struct
+            // TODO messages file
+            throw new DataException(String.format("Cannot generate schemas for mixed type arrays, consider using the %s SMT", ArrayFlatten.class.getName()));
```

Comment:
Obviously the user could still have conflicting schema definitions with a pipeline of "flatten -> map to struct" given two documents:
`widgets: [1,"foo"]`
`widgets: ["bar", 2]`
would result in schemas for `widgets.0` in the first case of int and in the second string.

Reply:
yes, but if a user is using mixed type arrays _and_ being inconsistent on the element ordering they have bigger problems than I'm willing to solve for them 😜